### ISSUE TITLE
Messy test output

### ISSUE THREAD

**Issue by kampelmuehler** [chenweikai/3PSDF#4]
Hi,
when using the provided example test command for single image reconstruction with the weights and data provided in the sharepoint, the resulting output meshes look beyond messy. Examples:
![grafik](https://user-images.githubusercontent.com/9306799/185363639-51f869d5-e55f-4244-ad79-5a77051cb55e.png)
![grafik](https://user-images.githubusercontent.com/9306799/185363902-c9da8302-402b-4113-a645-95af43e2642b.png)
![grafik](https://user-images.githubusercontent.com/9306799/185364869-c4faacbb-fb8c-4908-9476-00823f649289.png)

I could not find a single one that works. What do you reckon might be wrong here?
Thanks!

**Comment by chenweikai** [comment]
@clinplayer Could you help address this issue? Thanks!

**Comment by clinplayer** [comment]
As indicated in the details in the paper, we use post-processing steps to refine the results - flying parts cleaning, hole filling, and surface smoothing. 
I think the introduction of the NULL sign brings about the problem of uncontrollable incontinuity - the surface tends to be unsmooth and messy. Thus post-processings are recommended. 
Also, all the methods will have some examples that look pretty bad. For 3PSDF, it performs poorly in the watercraft and chair category. Could you please check out the "airplane" and "car" results to see if they are better?

**Comment by kampelmuehler** [comment]
Thanks for getting back to me @clinplayer 

Find here 3 cars and 3 airplanes. Are those expected outputs? Can you share your post processing steps somewhere?

![grafik](https://user-images.githubusercontent.com/9306799/185879354-e4c4b954-3be4-46ba-ba0b-46a3c1cc8165.png)
![grafik](https://user-images.githubusercontent.com/9306799/185879419-ffcef620-1793-4721-a736-c7ce5911d5a5.png)
![grafik](https://user-images.githubusercontent.com/9306799/185879469-ab1cfc9f-4f71-4075-8526-52fb36b12d9d.png)
![grafik](https://user-images.githubusercontent.com/9306799/185879633-de23fa11-196b-433d-9662-ea7adb0487a8.png)
![grafik](https://user-images.githubusercontent.com/9306799/185879733-2420a303-4756-44de-9094-af4bce88be8d.png)
![grafik](https://user-images.githubusercontent.com/9306799/185879799-e1e0c1af-91ed-420f-9e23-52871b3b8edb.png)

**Comment by clinplayer** [comment]
Hey, yes, the raw output seems reasonable. ShapeNet dataset contains various geometries within a category, and for some shapes with a limited amount, the network is struggling to learn good surfaces for SVR tasks... I have to admit this issue becomes worse for 3PSDF, since it does not have a closed and globally consistent (inside/outside) field initialization like SDF or occupancy.

We use the implementation of HC Laplacian smoothness and Close holes in Meshlab
https://github.com/cnr-isti-vclab/meshlab/tree/3c05b28ab83e0fd39ec732ca0aa629ef9475bdaf/src/meshlabplugins/filter_unsharp

After that, you can filter out the collapsed small components by measuring the length of the bbox. 

Some other tips that could help, from my experiences:
Try to use 3PSDF for shapes in a similar geometric distribution with less variety.
Some problems could be caused by convergence issues. You might try to retrain the network using the latest code (I fixed some small issues) and add a learning rate scheduler ( I didn't fine-tune the network exhaustively). 
It seems the normals are flipped. Try to flip them back before processing?

**Comment by kampelmuehler** [comment]
Ok, thanks for sharing this information :)
